File: src/Guioes/Testes_Exames/teste16.erl
-module(teste16).
-export([start/0, sinaliza/2, votos/3]).

start() ->
    spawn(fun() -> espera (0,0,0) end).

espera(valor, n1, n2) -> 
    if
        valor =:= 1 ->
            io:format("CHEGAMOS AO QUE É PRETENDIDO~n",[]);
        true -> 
            io:format("AINDA NÃO ESTÁ CERTO ~n",[])
    end,

    receive
        {sinaliza, a, b} ->
                espera(0, a, b);
        {T1,T2} ->
            Res = if
                T1 == n1 andalso T2 == n2 -> 1;
                true -> 0
            end,
        espera(Res,n1,n2)
end.

sinaliza(a, b) ->
    espera ! {sinaliza,a,b}.


votos(T1,T2,tipo) ->
    io:format("Sistema votos~n",[]),
    if 
        tipo =:= tipo1 ->
            espera ! {T1+1,T2};
        true ->
            espera ! {T1, T2+1}
    end.


   

File: src/Guioes/Testes_Exames/Teste16.java
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

class Teste16 extends Thread {
    public static void main(String[] args) {

        MeuMonitor evts = new MeuMonitor();

        new Thread (() -> {
            try{
                evts.espera(1, 4, 2, 2);
            } catch (Exception e) { }
        }).start();


        new Thread (() -> {
            try {
                System.out.println("vou sinalizar 1 1x");
                evts.sinaliza(1);
                Thread.sleep(3000);

                System.out.println("vou sinalizar 2 1x");
                evts.sinaliza(2);
                Thread.sleep(3000);

                System.out.println("vou sinalizar 1 2x");
                evts.sinaliza(1);
                Thread.sleep(3000);

                System.out.println("vou sinalizar 2 2x");
                evts.sinaliza(2);
                Thread.sleep(3000);

                System.out.println("vou sinalizar 1 3x");
                evts.sinaliza(1);
                Thread.sleep(3000);
                System.out.println("vou sinalizar 1 4x");
                evts.sinaliza(1);
                Thread.sleep(3000);


                
            } catch (Exception e ) { }
        }).start();

    }
}

class MeuMonitor {

    Lock lock = new ReentrantLock();
    Condition notThere = lock.newCondition();
    
    int tipo1;
    int n1;
    int tipo2;
    int n2;

    MeuMonitor() {
        tipo1 = 1;
        tipo2 = 2;
        n1 = 0;
        n2 = 0;
    }


    void sinaliza(int tipo) {
        lock.lock();
        try{
            if (tipo == tipo1) { n1 += 1; }
            if (tipo == tipo2) { n2 += 1; }
            notThere.signalAll();
        } finally {
            lock.unlock();
        }
    }
 
    void espera(int tipo1, int n1, int tipo2, int n2) throws InterruptedException {
        lock.lock();
        try{
            System.out.println("Vou esperar agora...");

            while (!(this.n1 == n1 && this.n2 == n2)) {
                System.out.println("Nao posso sair");
                notThere.await();
            }
            
            System.out.println("Saí da espera");
        } finally {
            lock.unlock();
        }
    }
}